### ISSUE TITLE
NSFS | S3 | Versioning: AccessDenied on GetObject in a version-enabled bucket despite bucket policy allowing

### ISSUE THREAD

**Issue by hseipp** [noobaa/noobaa-core#8368]
### Environment info
- NooBaa Version: **5.15.4**
- Platform: **CentOS 8, deployed using [https://github.com/IBM/StorageScaleVagrant](https://github.com/IBM/StorageScaleVagrant)**

### Actual behavior

GetObject returns AccessDenied although bucket policy allows access.

```
        policy_document = json.dumps(
        {
            "Version": "2012-10-17",
            "Statement": [{
                    "Effect": "Allow",
                    "Principal": { "AWS": "{}".format(get_alt_user_id()) },
                    "Action": [ "s3:GetObject", "s3:GetObjectVersion", "s3:ListBucketVersions" ],
                    "Resource": "arn:aws:s3:::{}/{}".format(bucket_name, key)
                }
            ]
        })

        client.put_bucket_policy(Bucket=bucket_name, Policy=policy_document)
    
>       response = alt_client.get_object(Bucket=bucket_name, Key=key)

s3tests_boto3/functional/test_s3.py:1759: 
...
E           botocore.exceptions.ClientError: An error occurred (AccessDenied) when calling the GetObject operation: Access Denied
```

### Expected behavior

GetObject should pass as policy has granted access to alternate user - here: CES S3 user with same gid as primary user, but different uid.

### Steps to reproduce

Run test_s3cever11 with Ceph s3-tests suite. Two users with the same gid are configured:

```
def test_s3cever11():
    client = get_client()
    bucket_name = get_new_bucket(client=client)

    # The following operation uses another client instance -> following operations will fail
#    check_configure_versioning_retry(bucket_name, "Enabled", "Enabled")
    _ces_safe_check_configure_versioning_retry(client, bucket_name, "Enabled", "Enabled")

    key = "test"
    version_ids = []
    contents = []
    num_versions = 3
    (version_ids, contents) = create_multiple_versions(client, bucket_name, key, num_versions)

    response = client.list_object_versions(Bucket=bucket_name)
    assert len(response['Versions']) == 3

    resource1 = "arn:aws:s3:::" + bucket_name
    resource2 = "arn:aws:s3:::" + bucket_name + "/*"
    policy_document = json.dumps(
    {
        "Version": "2012-10-17",
        "Statement": [{
        "Effect": "Allow",
        "Principal": {"AWS": "*"},
        "Action": "s3:ListBucketVersions",
        "Resource": [
            "{}".format(resource1),
            "{}".format(resource2)
          ]
        }]
     })

    client.put_bucket_policy(Bucket=bucket_name, Policy=policy_document)

    alt_client = get_alt_client()
    response = alt_client.list_object_versions(Bucket=bucket_name)
    assert response['ResponseMetadata']['HTTPStatusCode'] == 200

    response_code = ""
    try:
        response = alt_client.get_object(Bucket=bucket_name, Key=key, VersionId=version_ids[0])
    except ClientError as e:
        response_code = e.response['Error']['Code']

    assert response_code == "AccessDenied"

    # s3:versionid condition key is not supported by Noobaa, so skipping that policy
    policy_document = json.dumps(
    {
        "Version": "2012-10-17",
        "Statement": [{
                "Effect": "Allow",
                "Principal": { "AWS": "{}".format(get_alt_user_id()) },
                "Action": [ "s3:GetObject", "s3:GetObjectVersion" ],
                "Resource": "arn:aws:s3:::{}/{}".format(bucket_name, key)
            }
        ]
    })

    print(policy_document)
    client.put_bucket_policy(Bucket=bucket_name, Policy=policy_document)

    response = alt_client.get_object(Bucket=bucket_name, Key=key)
    body = _get_body(response)
    assert body == contents[2]

    response = alt_client.get_object(Bucket=bucket_name, Key=key, VersionId=version_ids[0])
    body = _get_body(response)
    assert body == contents[0]
    response = alt_client.get_object(Bucket=bucket_name, Key=key, VersionId=version_ids[1])
    body = _get_body(response)
    assert body == contents[1]
```

### More information - Screenshots / Logs / Other output

Interestingly when I execute a similar CLI sequence, I can't reproduce the error symptom:

```
s3 mb s3://shared2
s3api put-bucket-versioning --bucket shared2 --versioning-configuration '{"Status": "Enabled"}'
s3 cp ../custom-tests/foobar.txt s3://shared2/test
s3 cp ../custom-tests/foobarbaz.txt s3://shared2/test
# following fails as expected
s3u2 cp s3://shared2/test test
s3api put-bucket-policy --bucket shared2 --policy '{"Version": "2012-10-17","Statement": [{"Effect": "Allow","Principal": {"AWS": "s3user2"},"Action":["s3:GetObject", "s3:GetObjectVersion" ],"Resource": "arn:aws:s3:::shared2/test"}]}'
# following succeeds as expected
s3u2 cp s3://shared2/test test
s3u2api get-object --bucket shared2 --key test --version-id mtime-d47smoigl4ao-ino-kqo x
```

Log with "all" log level attached.
[noobaa_20240916_1.log.gz](https://github.com/user-attachments/files/17015245/noobaa_20240916_1.log.gz)

**Comment by shirady** [comment]
@hseipp  In the logs, I can see that:

> Sep 16 14:19:13 m1.example.com node[26226]: Sep-16 14:19:13.556 [nsfs/26301]    [L1] core.endpoint.s3.s3_bucket_policy_utils:: bucket_policy:  Action  fit? s3:ListBucketVersions s3:GetObjectVersion
> Sep 16 14:19:13 m1.example.com node[26226]: Sep-16 14:19:13.556 [nsfs/26301]    [L1] core.endpoint.s3.s3_bucket_policy_utils:: bucket_policy:  Principal  fit? SENSITIVE-7846cdd4c2b90527 s3user2
> Sep 16 14:19:13 m1.example.com node[26226]: Sep-16 14:19:13.556 [nsfs/26301]    [L1] core.endpoint.s3.s3_bucket_policy_utils:: bucket_policy:  Resource  fit? /^arn:aws:s3:::s3tests-cwd1j5qytc5i90iqt1wf3-1$/ arn:aws:s3:::s3tests-cwd1j5qytc5i90iqt1wf3-1/test
> Sep 16 14:19:13 m1.example.com node[26226]: Sep-16 14:19:13.556 [nsfs/26301]    [L1] core.endpoint.s3.s3_bucket_policy_utils:: bucket_policy:  Resource  fit? /^arn:aws:s3:::s3tests-cwd1j5qytc5i90iqt1wf3-1\/.*$/ arn:aws:s3:::s3tests-cwd1j5qytc5i90iqt1wf3-1/test
> Sep 16 14:19:13 m1.example.com node[26226]: Sep-16 14:19:13.556 [nsfs/26301]    [L1] core.endpoint.s3.s3_bucket_policy_utils:: bucket_policy: is_statements_fit false true true true
> Sep 16 14:19:13 m1.example.com node[26226]: Sep-16 14:19:13.556 [nsfs/26301] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>AccessDenied</Code><Message>Access Denied</Message><Resource>/s3tests-cwd1j5qytc5i90iqt1wf3-1/test?versionId=mtime-d47rvncf29kw-ino-d8g</Resource><RequestId>m153cob7-cwgief-18a4</RequestId></Error> GET /s3tests-cwd1j5qytc5i90iqt1wf3-1/test?versionId=mtime-d47rvncf29kw-ino-d8g {"host":"localhost:6443","accept-encoding":"identity","user-agent":"Boto3/1.35.4 md/Botocore#1.35.4 ua/2.0 os/linux#4.18.0-348.7.1.el8_5.x86_64 md/arch#x86_64 lang/python#3.8.8 md/pyimpl#CPython cfg/retry-mode#legacy Botocore/1.35.4","x-amz-date":"20240916T141913Z","x-amz-content-sha256":"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855","authorization":"AWS4-HMAC-SHA256 Credential=84zUfVLtmuPXVgqLt2qE/20240916/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=57d241b9e60c4990c9f62ea68ed76c89c9df099bc1267c9beb24c21bb9fd8d26","amz-sdk-invocation-id":"97c400f0-9e91-48d4-8556-3f718e0813e9","amz-sdk-request":"attempt=1"} Error: Access Denied

We iterate on all the actions and print if they fit, in this log, I can see only one check - `ListBucketVersions` and I would expect to see also `s3:GetObject` and `3:GetObjectVersion` in the printing:
https://github.com/noobaa/noobaa-core/blob/b28bd87e37e825456e4f7b6c79366a2fb123b3bd/src/endpoint/s3/s3_bucket_policy_utils.js#L152-L162

Could you please add to your script (that appears under the "Actual behavior" title) printing of the bucket policy as it was saved (print the get_bucket_policy)? and also the printing of `"AWS": "{}".format(get_alt_user_id())` just to be sure the alternative account is set as needed in the configuration and attach the collected logs.

**Comment by hseipp** [comment]
Please note that in the meantime my test system got upgraded to noobaa-core 5.17.0-20240904.
I can still re-produce the issue.
So I changed the test code to
```
def test_s3cever11():
    client = get_client()
    bucket_name = get_new_bucket(client=client)

    # The following operation uses another client instance -> following operations will fail
#    check_configure_versioning_retry(bucket_name, "Enabled", "Enabled")
    _ces_safe_check_configure_versioning_retry(client, bucket_name, "Enabled", "Enabled")

    key = "test"
    version_ids = []
    contents = []
    num_versions = 3
    (version_ids, contents) = create_multiple_versions(client, bucket_name, key, num_versions)

    response = client.list_object_versions(Bucket=bucket_name)
    assert len(response['Versions']) == 3

    resource1 = "arn:aws:s3:::" + bucket_name
    resource2 = "arn:aws:s3:::" + bucket_name + "/*"
    policy_document = json.dumps(
    {
        "Version": "2012-10-17",
        "Statement": [{
        "Effect": "Allow",
        "Principal": {"AWS": "*"},
        "Action": "s3:ListBucketVersions",
        "Resource": [
            "{}".format(resource1),
            "{}".format(resource2)
          ]
        }]
     })

    client.put_bucket_policy(Bucket=bucket_name, Policy=policy_document)

    alt_client = get_alt_client()
    response = alt_client.list_object_versions(Bucket=bucket_name)
    assert response['ResponseMetadata']['HTTPStatusCode'] == 200

    response_code = ""
    try:
        response = alt_client.get_object(Bucket=bucket_name, Key=key, VersionId=version_ids[0])
    except ClientError as e:
        response_code = e.response['Error']['Code']

    assert response_code == "AccessDenied"

    # s3:versionid condition key is not supported by Noobaa, so skipping that policy
    policy_document = json.dumps(
    {
        "Version": "2012-10-17",
        "Statement": [{
                "Effect": "Allow",
                "Principal": { "AWS": "{}".format(get_alt_user_id()) },
                "Action": [ "s3:GetObject", "s3:GetObjectVersion" ],
                "Resource": "arn:aws:s3:::{}/{}".format(bucket_name, key)
            },
        ]
    })

    print(policy_document)
    client.put_bucket_policy(Bucket=bucket_name, Policy=policy_document)

    response = client.get_bucket_policy(Bucket=bucket_name, ExpectedBucketOwner='s3user1')
    print(response['Policy'])

    response = alt_client.get_object(Bucket=bucket_name, Key=key)
    body = _get_body(response)
    assert body == contents[2]

    response = alt_client.get_object(Bucket=bucket_name, Key=key, VersionId=version_ids[0])
    body = _get_body(response)
    assert body == contents[0]
    response = alt_client.get_object(Bucket=bucket_name, Key=key, VersionId=version_ids[1])
    body = _get_body(response)
    assert body == contents[1]
```

and get the same test result as above

```
>       response = alt_client.get_object(Bucket=bucket_name, Key=key)

s3tests_boto3/functional/test_s3.py:1761: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox/py/lib/python3.8/site-packages/botocore/client.py:569: in _api_call
    return self._make_api_call(operation_name, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <botocore.client.S3 object at 0x7f56bb155970>, operation_name = 'GetObject', api_params = {'Bucket': 's3tests-o2061vmb4d278hurumavk-1', 'Key': 'test'}
...
E           botocore.exceptions.ClientError: An error occurred (AccessDenied) when calling the GetObject operation: Access Denied
```

 plus the following output for the print statements:

```
{"Version": "2012-10-17", "Statement": [{"Effect": "Allow", "Principal": {"AWS": "s3user2"}, "Action": ["s3:GetObject", "s3:GetObjectVersion"], "Resource": "arn:aws:s3:::s3tests-o2061vmb4d278hurumavk-1/test"}]}
{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"AWS":["s3user2"]},"Action":["s3:GetObject","s3:GetObjectVersion"],"Resource":"arn:aws:s3:::s3tests-o2061vmb4d278hurumavk-1/test"}]}
```

Relevant sections of my `s3tests.conf`:
```
[s3 main]
# main display_name set in vstart.sh
display_name = s3user

# main user_idname set in vstart.sh
user_id = s3user1
...
[s3 alt]
# alt display_name set in vstart.sh
display_name = s3user2
## alt email set in vstart.sh
email = s3user2@example.com

# alt user_id set in vstart.sh
user_id = s3user2
...

Noobaa config:
```
$ sudo cat /ibm/cesShared/ces/s3-config/accounts/s3user1.json | jq '.name, .email'
"s3user1"
"s3user1"
$ sudo cat /ibm/cesShared/ces/s3-config/accounts/s3user2.json | jq '.name, .email'
"s3user2"
"s3user2"
```

The corresponding "all" debug level log is attached
[noobaa_20240918_1.log.gz](https://github.com/user-attachments/files/17040550/noobaa_20240918_1.log.gz)

**Comment by shirady** [comment]
I'm closing the issue, if needed @hseipp feel free to re-open.
As mentioned, I opened another issue for the bucket cache #8391.
cc: @nadavMiz, @romayalon

**Comment by hseipp** [comment]
> I will open a different issue for the `bucket_namespace_cache` updates between forks. @hseipp would you please validate that with the `time.sleep(65)` it works for you?
> 
> cc: @nadavMiz @romayalon

I can confirm that with the `time.sleep(65)` delay the test passes on my test environment.

**Comment by shirady** [comment]
Hi @hseipp,
I want to let you know that we merged a code change in PR #8527.
Could you please test it again on the master branch? (on RPM that includes the code changes)

In case it didn't solve the issue, please attach all the details:
- RPM version and date
- Partial code snippet (to see that it is with the code change)
- The steps that you did
- Logs with high debug level

I'm adding a label for validation, please update in a comment the status.
